babble/src/main/java/io/mosaicnetworks/babble/servicediscovery/webrtc/WebRTCService.java

```diff
@@ -208,6 +253,44 @@ public void onErrorResponse(VolleyError error) {
 
     @Override
     public void stopAdvertising() {
-        //do nothing
+        Disco group = ConfigManager.getInstance(null).getDisco();
```

Comment:
Is this safe? ConfigManager is a static class created by the fragments on MainActivity. You cannot assume that it persists without being garbage collected into the GameActivity lifecycle. This feels potentially like one of those oddly persistent intermittent bugs. 

It probably requires storing the Disco instance on WebRTCService itself. 

What do you think @joneskm?

Reply:
Yes I was confused by the ConfigManager class. Not sure where else to get the "current" Disco group from. Either we keep a variable in WebRTCService to track the "current" group, or we add a groupID parameter to stopAdvertising

Reply:
In my major rework branch, the Disco object is stored on the WebRTCService which polls the disco server every, say, 5 minutes to the update endpoint. This has the added benefit that the Disco Server can implement a simple timeout of say 10 minutes to remove groups. 

This means the groups tidy themselves even in the event of a crash. 

You could rewrite the condition for deletion to be if leaving if the only peer. 

And because it is polling - including refreshing the peers in the disco group - it makes sense to keep a copy on WebRTCService.

You can also hook update on join / leave requests too...

Reply:
I would keep that for a different pull request.

Reply:
Yes, but I was providing the detail for context. In this PR I would create mDisco on WebRTCService and populate it. 

All the rest of it is a potential enhancement that would build on it in a later PR.

Reply:
That's not a solution either.

Ultimately, all the info needed to advertise a group should come from the function parameters, which are currently not used at all. 

```
   public boolean advertise(String genesisPeers, String currentPeers, PeersProvider peersProvider) {
        ConfigManager configManager = ConfigManager.getInstance(null);
        Disco disco = configManager.getDisco();
        sendGroupToDisco(REGISTER_END_POINT, disco);
        return true;
    }
```

Whether this functions gets Disco from WebRTCService or ConfigManager, is wrong either way. It should be a parameter passed by the function caller, which is BabbleService `start`.

This is evidence that the ServiceAdvertiser interface needs to be changed. The same goes for `stopAdvertising`, which should take a group ID.

I would change the signature of the `advertise` method to accept a `Disco` object, but it seems that Disco is defined in the WebRTC package so not suitable for the mDNS service discovery.

So I think this is part of a much bigger problem that should be addressed separately. Persitent intermittent bugs will remind us that the problem exists.

Reply:
```java
    private Disco mDisco;

...

    public WebRTCAdvertiser(Context context, Disco disco) {
        Log.i(TAG, "WebRTCAdvertiser: Created");
        mDisco = disco;
        mServiceName = disco.GroupUID;
        mQueue = Volley.newRequestQueue(context.getApplicationContext());
    }
```

Is the code in revised version.

It is fudged that only a new group advertises in WebRTC (we can address that later), but the service start is as follows:

```java
    @Override
    protected void onServiceConnected() {
        try {
            mBoundService.setPublicKey(mPubKeyHex);
            mBoundService.start(mConfigDirectory, mGroupDescriptor, mServiceAdvertiser);
```


Which brings me to realise the major difference that I had forgotten...

I split the WebRTCService and actually has a WebRTCAdvertiser class that sat more sensibly.

https://github.com/mosaicnetworks/babble-android/blob/major-refactor/babble/src/main/java/io/mosaicnetworks/babble/servicediscovery/webrtc/WebRTCAdvertiser.java

Reply:
yeah I saw that. It matches what exists for the mdnsAdvertiser. A friday afternoon whiteboard session would be good here...